### PR TITLE
Add height for colpolygon

### PR DESCRIPTION
**Pull request by CrosRoad95**
Functions to set/get height of colpolygon. "false" as argument resets to +- float min, max value ( basicly, infinity ). Both are client and server side.
```lua
setColPolygonHeight( colpolygon, floor, roof ) 
polygon:setHeight(floor, roof)

floor, roof = getColPolygonHeight( colpolygon )
floor, roof = colpolygon:getHeight()
```

I don't know what version i need to set in Server/mods/deathmatch/logic/CResourceChecker.Data.h, i set 1.5.7 for now.

```cpp
        {"setColPolygonHeight", "1.5.7"},
        {"setColPolygonHeight", "1.5.7"},
```

test resource: [polygonheight.zip](https://github.com/multitheftauto/mtasa-blue/files/5654027/polygonheight.zip)

### CODE REVIEW

File: Server/mods/deathmatch/logic/packets/CEntityAddPacket.cpp [hunk 1]
```diff
@@ -1025,6 +1025,15 @@ bool CEntityAddPacket::Write(NetBitStreamInterface& BitStream) const
                                 vertex.data.vecPosition = *iter;
                                 BitStream.Write(&vertex);
                             }
+
+                            if (BitStream.Version() >= 0x1DA)
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/packets/CEntityAddPacket.cpp]: Don't use BitStream.Version() use `BitStream.Can()`

File: Server/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.h [hunk 2]
```diff
@@ -38,4 +40,7 @@ class CLuaColShapeDefs : public CLuaDefs
 
     LUA_DECLARE(IsInsideColShape);
     LUA_DECLARE(GetColShapeType);
+
+	static std::tuple<float, float> GetColPolygonHeight(CColPolygon* pColPolygon);
+    static bool                     SetColPolygonHeight(CColPolygon* pColPolygon, std::variant<bool, float> floor, std::variant<bool, float> ceil);
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.h]: formatting

File: Server/mods/deathmatch/logic/CResourceChecker.Data.h [hunk 3]
```diff
@@ -234,6 +234,8 @@ namespace
         {"getVehicleModelWheelSize", "1.5.7-9.20642"},
         {"setVehicleWheelScale", "1.5.7-9.20642"},
         {"getVehicleWheelScale", "1.5.7-9.20642"},
+        {"setColPolygonHeight", "1.5.7"},
+        {"setColPolygonHeight", "1.5.7"},
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CResourceChecker.Data.h]: we're on 1.5.8

File: Client/mods/deathmatch/logic/rpc/CColShapeRPCs.h [hunk 4]
```diff
@@ -23,4 +23,5 @@ class CColShapeRPCs : public CRPCFunctions
     DECLARE_ELEMENT_RPC(UpdateColPolygonPoint);
     DECLARE_ELEMENT_RPC(AddColPolygonPoint);
     DECLARE_ELEMENT_RPC(RemoveColPolygonPoint);
-};
+    DECLARE_ELEMENT_RPC(SetColShapePolygonHeight);
+};
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/rpc/CColShapeRPCs.h]: Missing newline at end of file

File: Client/mods/deathmatch/logic/rpc/CColShapeRPCs.cpp [hunk 5]
```diff
@@ -79,3 +80,14 @@ void CColShapeRPCs::RemoveColPolygonPoint(CClientEntity* pSource, NetBitStreamIn
...
+void CColShapeRPCs::SetColShapePolygonHeight(CClientEntity* pSource, NetBitStreamInterface& bitStream)
+{
+    float         fFloor;
+    float         fCeil;
+    if (bitStream.Read(fFloor) && bitStream.Read(fCeil))
+    {
+        CClientColPolygon* pColShape = static_cast<CClientColPolygon*>(pSource);
+        CStaticFunctionDefinitions::SetColPolygonHeight(pColShape, fFloor, fCeil);
+    }
+}
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/rpc/CColShapeRPCs.cpp]: Incorrectly formatted

File: Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp [hunk 6]
```diff
@@ -7377,6 +7377,19 @@ bool CStaticFunctionDefinitions::RemoveColPolygonPoint(CClientColPolygon* pColPo
...
+    if (pColPolygon->SetHeight(fFloor, fCeil))
+    {
+        RefreshColShapeColliders(pColPolygon);
+        return true;
+    }
+
+    return false;
+}
+
+
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]: Don't add to CStaticFunctionDefinitions when using the new parser, just add code directly to lua definitions

**Review by qaisjp — CHANGES_REQUESTED** [review]

File: Client/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.h [hunk 7]
```diff
@@ -11,6 +11,7 @@
 
 #pragma once
 #include "CLuaDefs.h"
+#include "lua/CLuaFunctionParser.h"
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.h]: Add this line to .cpp not to .h

File: Client/mods/deathmatch/logic/CClientColPolygon.cpp [hunk 8]
```diff
@@ -182,14 +198,17 @@ void CClientColPolygon::DebugRender(const CVector& vecPosition, float fDrawRadiu
         {
             float fZ = vecPosition.fZ - fDrawRadius + fDrawRadius * 2.0f * (s / (float)(uiNumSlices - 1));
             fZ += 4;            // Extra bit so a slice is on the same Z coord as the camera
-            for (uint i = 0; i < uiNumPoints; i++)
+            if (m_fFloor < fZ && fZ < m_fCeil)
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CClientColPolygon.cpp]: shouldn't this be inclusive?

**Comment by CrosRoad95** [comment]
Fixed, and i'm thinking it is ready to marge

File: Client/mods/deathmatch/logic/CClientColPolygon.h [hunk 9]
```diff
@@ -34,6 +34,8 @@ class CClientColPolygon : public CClientColShape
     unsigned int                           CountPoints() const { return static_cast<unsigned int>(m_Points.size()); };
     std::vector<CVector2D>::const_iterator IterBegin() { return m_Points.begin(); };
     std::vector<CVector2D>::const_iterator IterEnd() { return m_Points.end(); };
+    bool SetHeight(float fFloor, float fCeil);
+    void GetHeight(float &fFloor, float &fCeil) { fFloor = m_fFloor; fCeil = m_fCeil; };
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/CClientColPolygon.h]: - Nit: ampersand should be together with `float`
- Declare as `const`
```suggestion
    void GetHeight(float& fFloor, float& fCeil) const { fFloor = m_fFloor; fCeil = m_fCeil; };
```

File: Client/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.cpp [hunk 10]
```diff
@@ -810,3 +815,38 @@ int CLuaColShapeDefs::RemoveColPolygonPoint(lua_State* luaVM)
     argStream.SetCustomError("ColShape must be Polygon");
     return luaL_error(luaVM, argStream.GetFullErrorMessage());
 }
+
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.cpp:818]: Nit: unwanted extra new line.
```suggestion
```

File: Client/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.h [hunk 11]
```diff
@@ -37,4 +38,7 @@ class CLuaColShapeDefs : public CLuaDefs
 
     LUA_DECLARE(IsInsideColShape);
     LUA_DECLARE(GetColShapeType);
+
+    static std::tuple<float, float> GetColPolygonHeight(CClientColPolygon* pColPolygon);
+    static bool SetColPolygonHeight(CClientColPolygon* pColPolygon, std::variant<bool, float> floor, std::variant<bool, float> ceil);
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.h]: Nit: fix formatting
```suggestion
    static bool                     SetColPolygonHeight(CClientColPolygon* pColPolygon, std::variant<bool, float> floor, std::variant<bool, float> ceil);
```

File: Client/mods/deathmatch/logic/rpc/CColShapeRPCs.h [hunk 12]
```diff
@@ -23,4 +23,6 @@ class CColShapeRPCs : public CRPCFunctions
     DECLARE_ELEMENT_RPC(UpdateColPolygonPoint);
     DECLARE_ELEMENT_RPC(AddColPolygonPoint);
     DECLARE_ELEMENT_RPC(RemoveColPolygonPoint);
+    DECLARE_ELEMENT_RPC(SetColShapePolygonHeight);
+
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/rpc/CColShapeRPCs.h]: Nit: unwanted extra new line.
```suggestion
```

File: Client/mods/deathmatch/logic/CPacketHandler.cpp [hunk 13]
```diff
@@ -3964,6 +3964,13 @@ void CPacketHandler::Packet_EntityAdd(NetBitStreamInterface& bitStream)
                                 pPolygon->AddPoint(vertex.data.vecPosition);
                             }
                             pEntity = pShape = pPolygon;
+                            if (bitStream.Can(eBitStreamVersion::SetPolygonHeight))
+                            {
+                                float fFloor, fCeil;
+                                bitStream.Read(fFloor);
+                                bitStream.Read(fCeil);
+                                pPolygon->SetHeight(fFloor, fCeil);
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/CPacketHandler.cpp]:
```suggestion
                                if (bitStream.Read(fFloor) && bitStream.Read(fCeil))
                                    pPolygon->SetHeight(fFloor, fCeil);
```

File: Server/mods/deathmatch/logic/CColPolygon.h [hunk 14]
```diff
@@ -38,6 +38,9 @@ class CColPolygon : public CColShape
     std::vector<CVector2D>::const_iterator IterBegin() { return m_Points.begin(); };
     std::vector<CVector2D>::const_iterator IterEnd() { return m_Points.end(); };
 
+    bool SetHeight(float fFloor, float fCeil);
+    void GetHeight(float &fFloor, float &fCeil) { fFloor = m_fFloor; fCeil = m_fCeil; };
```
Comment (patrikjuvonen) [inline Server/mods/deathmatch/logic/CColPolygon.h]: - Nit: ampersand should be together with `float`
- Declare as `const`

```suggestion
    void GetHeight(float& fFloor, float& fCeil) const { fFloor = m_fFloor; fCeil = m_fCeil; };
```

File: Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp [hunk 15]
```diff
@@ -12182,3 +12182,19 @@ CMtaVersion CStaticFunctionDefinitions::GetVersionSortable()
...
     return SString("%d.%d.%d-%d.%05d.%d", MTASA_VERSION_MAJOR, MTASA_VERSION_MINOR, MTASA_VERSION_MAINTENANCE, MTASA_VERSION_TYPE, MTASA_VERSION_BUILD, 0);
 }
+
+bool CStaticFunctionDefinitions::SetColPolygonHeight(CColPolygon* pColPolygon, float fFloor, float fCeil)
+{
+    if (pColPolygon->SetHeight(fFloor, fCeil))
+    {
+        RefreshColShapeColliders(pColPolygon);
+
+        CBitStream      BitStream;
```
Comment (patrikjuvonen) [inline Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]: Nit: fix formatting
```suggestion
        CBitStream BitStream;
```

File: Shared/sdk/net/rpc_enums.h [hunk 16]
```diff
@@ -271,6 +271,7 @@ enum eElementRPCFunctions
   
     SET_DISCORD_JOIN_PARAMETERS,
     SET_TRANSFER_BOX_VISIBILITY,
+    COLSHAPE_POLYGON_SET_HEIGHT,
```
Comment (patrikjuvonen) [inline Shared/sdk/net/rpc_enums.h]: Perhaps rename to `SET_COLPOLYGON_HEIGHT` for consistency.
```suggestion
    SET_COLPOLYGON_HEIGHT,
```

File: Client/mods/deathmatch/logic/rpc/CColShapeRPCs.cpp [hunk 17]
```diff
@@ -19,6 +19,7 @@ void CColShapeRPCs::LoadFunctions(void)
     AddHandler(UPDATE_COLPOLYGON_POINT, UpdateColPolygonPoint, "UpdateColPolygonPoint");
     AddHandler(ADD_COLPOLYGON_POINT, AddColPolygonPoint, "AddColPolygonPoint");
     AddHandler(REMOVE_COLPOLYGON_POINT, RemoveColPolygonPoint, "RemoveColPolygonPoint");
+    AddHandler(COLSHAPE_POLYGON_SET_HEIGHT, SetColShapePolygonHeight, "SetColShapePolygonHeight");
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/rpc/CColShapeRPCs.cpp]: Perhaps rename to `SET_COLPOLYGON_HEIGHT` for consistency.
```suggestion
    AddHandler(SET_COLPOLYGON_HEIGHT, SetColShapePolygonHeight, "SetColShapePolygonHeight");
```

File: Client/mods/deathmatch/logic/rpc/CElementRPCs.cpp [hunk 18]
```diff
@@ -48,6 +48,7 @@ void CElementRPCs::LoadFunctions()
     AddHandler(SET_WEAPON_OWNER, SetWeaponOwner, "setWeaponOwner");
     AddHandler(SET_CUSTOM_WEAPON_FLAGS, SetWeaponConfig, "setWeaponFlags");
     AddHandler(SET_PROPAGATE_CALLS_ENABLED, SetCallPropagationEnabled, "setCallPropagationEnabled");
+    AddHandler(COLSHAPE_POLYGON_SET_HEIGHT, SetColPolygonHeight, "setColShapePolygonHeight");
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/rpc/CElementRPCs.cpp]: Perhaps rename to `SET_COLPOLYGON_HEIGHT` for consistency.
```suggestion
    AddHandler(SET_COLPOLYGON_HEIGHT, SetColPolygonHeight, "setColShapePolygonHeight");
```

File: Client/mods/deathmatch/logic/CClientColPolygon.cpp [hunk 19]
```diff
@@ -28,6 +28,11 @@ bool CClientColPolygon::DoHitDetection(const CVector& vecNowPosition, float fRad
     if (!IsInBounds(vecNowPosition))
         return false;
 
+    float z = vecNowPosition.fZ;
+
+    if (z > m_fCeil || z < m_fFloor)
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/CClientColPolygon.cpp]: Unnecessary variable.
```suggestion
    if (vecNowPosition.fZ > m_fCeil || vecNowPosition.fZ < m_fFloor)
```

File: Server/mods/deathmatch/logic/CColPolygon.cpp [hunk 20]
```diff
@@ -34,6 +34,11 @@ bool CColPolygon::DoHitDetection(const CVector& vecNowPosition)
     if (!IsInBounds(vecNowPosition))
         return false;
 
+    float z = vecNowPosition.fZ;
+
+    if (z > m_fCeil || z < m_fFloor)
```
Comment (patrikjuvonen) [inline Server/mods/deathmatch/logic/CColPolygon.cpp]: Unnecessary variable.
```suggestion
    if (vecNowPosition.fZ > m_fCeil || vecNowPosition.fZ < m_fFloor)
```

File: Shared/sdk/net/bitstream.h [hunk 21]
```diff
@@ -452,6 +452,8 @@ enum class eBitStreamVersion : unsigned short
     // 2020-11-10 0x71
     PedEnterExit,
 
+    SetPolygonHeight,
```
Comment (patrikjuvonen) [inline Shared/sdk/net/bitstream.h]: Better clarification.
```suggestion
    SetColPolygonHeight,
```

File: Client/mods/deathmatch/logic/CPacketHandler.cpp [hunk 22]
```diff
@@ -3964,6 +3964,13 @@ void CPacketHandler::Packet_EntityAdd(NetBitStreamInterface& bitStream)
                                 pPolygon->AddPoint(vertex.data.vecPosition);
                             }
                             pEntity = pShape = pPolygon;
+                            if (bitStream.Can(eBitStreamVersion::SetPolygonHeight))
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/CPacketHandler.cpp]: Better clarification.
```suggestion
                            if (bitStream.Can(eBitStreamVersion::SetColPolygonHeight))
```

File: Server/mods/deathmatch/logic/packets/CEntityAddPacket.cpp [hunk 23]
```diff
@@ -1025,6 +1025,15 @@ bool CEntityAddPacket::Write(NetBitStreamInterface& BitStream) const
                                 vertex.data.vecPosition = *iter;
                                 BitStream.Write(&vertex);
                             }
+
+                            if (BitStream.Can(eBitStreamVersion::SetPolygonHeight))
```
Comment (patrikjuvonen) [inline Server/mods/deathmatch/logic/packets/CEntityAddPacket.cpp]: Better clarification.
```suggestion
                            if (BitStream.Can(eBitStreamVersion::SetColPolygonHeight))
```

File: Server/mods/deathmatch/logic/CStaticFunctionDefinitions.h [hunk 24]
```diff
@@ -466,6 +466,7 @@ class CStaticFunctionDefinitions
     static bool AddColPolygonPoint(CColPolygon* pColPolygon, const CVector2D& vecPoint);
     static bool AddColPolygonPoint(CColPolygon* pColPolygon, uint uiPointIndex, const CVector2D& vecPoint);
     static bool RemoveColPolygonPoint(CColPolygon* pColPolygon, uint uiPointIndex);
+    static bool SetColPolygonHeight(CColPolygon* pPolygon, float fFloor, float fCeil);
```
Comment (patrikjuvonen) [inline Server/mods/deathmatch/logic/CStaticFunctionDefinitions.h]: Nit: naming consistency
```suggestion
    static bool SetColPolygonHeight(CColPolygon* pColPolygon, float fFloor, float fCeil);
```

File: Server/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.h [hunk 25]
```diff
@@ -38,4 +40,7 @@ class CLuaColShapeDefs : public CLuaDefs
 
     LUA_DECLARE(IsInsideColShape);
     LUA_DECLARE(GetColShapeType);
+
+	static std::tuple<float, float> GetColPolygonHeight(CColPolygon* pColPolygon);
+	static bool                     SetColPolygonHeight(CColPolygon* pColPolygon, std::variant<bool, float> floor, std::variant<bool, float> ceil);
```
Comment (patrikjuvonen) [inline Server/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.h]: Fix indentation.
```suggestion
    static std::tuple<float, float> GetColPolygonHeight(CColPolygon* pColPolygon);
    static bool                     SetColPolygonHeight(CColPolygon* pColPolygon, std::variant<bool, float> floor, std::variant<bool, float> ceil);
```

File: Server/mods/deathmatch/logic/CPerfStat.RPCPacketUsage.cpp [hunk 26]
```diff
@@ -219,6 +219,7 @@ ADD_ENUM1(ADD_COLPOLYGON_POINT)
 ADD_ENUM1(REMOVE_COLPOLYGON_POINT)
 ADD_ENUM1(UPDATE_COLPOLYGON_POINT)
 ADD_ENUM1(SET_DISCORD_JOIN_PARAMETERS)
+ADD_ENUM1(COLSHAPE_POLYGON_SET_HEIGHT)
```
Comment (patrikjuvonen) [inline Server/mods/deathmatch/logic/CPerfStat.RPCPacketUsage.cpp]: Perhaps rename to `SET_COLPOLYGON_HEIGHT` for consistency.
```suggestion
ADD_ENUM1(SET_COLPOLYGON_HEIGHT)
```

File: Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp [hunk 27]
```diff
@@ -12182,3 +12182,19 @@ CMtaVersion CStaticFunctionDefinitions::GetVersionSortable()
...
+        BitStream.pBitStream->Write(fFloor);
+        BitStream.pBitStream->Write(fCeil);
+        m_pPlayerManager->BroadcastOnlyJoined(CElementRPCPacket(pColPolygon, COLSHAPE_POLYGON_SET_HEIGHT, *BitStream.pBitStream));
```
Comment (patrikjuvonen) [inline Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]: Perhaps rename to `SET_COLPOLYGON_HEIGHT` for consistency.
```suggestion
        m_pPlayerManager->BroadcastOnlyJoined(CElementRPCPacket(pColPolygon, SET_COLPOLYGON_HEIGHT, *BitStream.pBitStream));
```

File: Client/mods/deathmatch/logic/CClientColPolygon.cpp [hunk 28]
```diff
@@ -200,9 +219,31 @@ void CClientColPolygon::DebugRender(const CVector& vecPosition, float fDrawRadiu
...
+    for (uint i = 0; i < uiNumPoints; i++)
+    {
+        const CVector2D& vecPointBegin = m_Points[i];
+        const CVector2D& vecPointEnd = m_Points[(i + 1) % uiNumPoints];
+
+        CVector vecBegin(vecPointBegin.fX, vecPointBegin.fY, m_fCeil);
+        CVector vecEnd(vecPointEnd.fX, vecPointEnd.fY, m_fCeil);
+        pGraphics->DrawLine3DQueued(vecBegin, vecEnd, fLineWidth, color, false);
+    }
+
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/CClientColPolygon.cpp]: Nit: unwanted extra new line.
```suggestion
```

File: Client/mods/deathmatch/logic/CClientColPolygon.cpp [hunk 29]
```diff
@@ -200,9 +219,31 @@ void CClientColPolygon::DebugRender(const CVector& vecPosition, float fDrawRadiu
...
+    for (uint i = 0; i < uiNumPoints; i++)
+    {
+        const CVector2D& vecPointBegin = m_Points[i];
+        const CVector2D& vecPointEnd = m_Points[(i + 1) % uiNumPoints];
+
+        CVector vecBegin(vecPointBegin.fX, vecPointBegin.fY, m_fCeil);
+        CVector vecEnd(vecPointEnd.fX, vecPointEnd.fY, m_fCeil);
+        pGraphics->DrawLine3DQueued(vecBegin, vecEnd, fLineWidth, color, false);
+    }
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/CClientColPolygon.cpp]: Can these be merged in any way?

**Review by patrikjuvonen — CHANGES_REQUESTED** [review]
Thanks for the pull request. I've reviewed some formatting issues. All seems to be working 
good according to your provided test resource.

**Review by patrikjuvonen — APPROVED** [review]
Thanks for the pull request. Tested working with the provided test resource.

**Pull request merged**